[ MacOS Debug ] fast/text/emoji-num-glyphs.html is a constant timeout

anttijk opened the pull request:
#### 228915f55f2f7b0def26f44c01d118a163f6fd8b
<pre>
[ MacOS Debug ] fast/text/emoji-num-glyphs.html is a constant timeout
<a href="https://bugs.webkit.org/show_bug.cgi?id=270195">https://bugs.webkit.org/show_bug.cgi?id=270195</a>
<a href="https://rdar.apple.com/problem/123720564">rdar://problem/123720564</a>

Reviewed by Alan Baradlay.

O(n^2) in whitespace renderer elimination code.

* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateBeforeDescendants):

Also store ::before as a previous renderer.

(WebCore::RenderTreeUpdater::textRendererIsNeeded):

Track whether we have seen an in-flow renderer already.
We can use this to determine if we need to create a whitespace renderer or not.

(WebCore::RenderTreeUpdater::storePreviousRenderer):

Remember in parent stack whether we have a preceding in-flow sibling.

* Source/WebCore/rendering/updating/RenderTreeUpdater.h:

Canonical link: <a href="https://commits.webkit.org/279648@main">https://commits.webkit.org/279648@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/46bbc6a36d70fff53a7338de388c6a70f62c6820

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54078 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33457 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6610 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57353 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4802 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56381 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40975 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4695 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43791 "Passed tests") | [❌ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3188 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56174 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31692 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46811 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24933 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28510 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4127 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2951 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/50214 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4330 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58947 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29270 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/4433 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51207 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30448 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46927 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50566 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11786 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31409 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30230 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on LayoutTests/platform/mac/TestExpectations:
```diff
@@ -2506,7 +2506,7 @@ webkit.org/b/269873 imported/w3c/web-platform-tests/webrtc-extensions/transfer-d
 [ Sonoma ] svg/as-image/img-with-svg-resource-not-in-dom-no-size-and-drawImage.html [ ImageOnlyFailure ]
 
-webkit.org/b/270195 [ Debug ] fast/text/emoji-num-glyphs.html [ Timeout ]
+fast/text/emoji-num-glyphs.html [ Pass ]
```
should this just be removed?

anttijk replied:
It is disabled in the main TestExpectations. It requires some emoji support that maybe not all platforms have.

anttijk committed:
[ MacOS Debug ] fast/text/emoji-num-glyphs.html is a constant timeout
https://bugs.webkit.org/show_bug.cgi?id=270195
rdar://problem/123720564

Reviewed by Alan Baradlay.

O(n^2) in whitespace renderer elimination code.

* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateBeforeDescendants):

Also store ::before as a previous renderer.

(WebCore::RenderTreeUpdater::textRendererIsNeeded):

Track whether we have seen an in-flow renderer already.
We can use this to determine if we need to create a whitespace renderer or not.

(WebCore::RenderTreeUpdater::storePreviousRenderer):

Remember in parent stack whether we have a preceding in-flow sibling.

* Source/WebCore/rendering/updating/RenderTreeUpdater.h:

Canonical link: https://commits.webkit.org/279648@main
